chore: redis cache for calendars in api v2

ThyMinimalDev opened the pull request:
## What does this PR do?

This PR implements Redis caching for calendar data in API v2 and refactors the caching architecture based on code review feedback. The main changes include:

- **Created dedicated `CalendarsCacheService`** - Extracted Redis caching operations into a separate service with proper dependency injection
- **Refactored dependency injection** - Updated `CalendarsController` to directly inject and use `CalendarsCacheService` instead of delegating through `CalendarsService`
- **Improved cache invalidation** - Added cache deletion calls across multiple services when calendar data changes
- **Enhanced code quality** - Extracted TTL constant and removed manual JSON serialization, letting `RedisService` handle it

This addresses feedback from @supalarry to eliminate unnecessary service delegation and improve separation of concerns.

**Link to Devin run**: https://app.devin.ai/sessions/856e14d526f7448da60fd0c28e6f1f83  
**Requested by**: morgan@cal.com

## How should this be tested?

- **Environment**: Redis should be available for caching functionality
- **Test scenarios**:
  1. Verify calendar data is cached on first API call and served from cache on subsequent calls
  2. Test cache invalidation when calendar credentials are deleted
  3. Confirm cache invalidation when selected calendars or destination calendars are modified
  4. Verify cache TTL (10 seconds) expires properly

**⚠️ Critical Review Areas:**
- **Method references**: Verify all calls to calendar cache methods work correctly after the refactoring
- **Cache invalidation coverage**: Ensure all calendar data modifications properly clear cache
- **Dependency injection**: Confirm RedisModule is imported in all necessary modules

## Mandatory Tasks (DO NOT REMOVE)


- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox. **N/A** - Internal caching implementation doesn't require user-facing documentation changes.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## Checklist

- My code follows the style guidelines of this project
- I have checked if my changes generate no new warnings
- I have verified all cache invalidation points are covered
- I have confirmed dependency injection setup is complete across all affected modules

ThyMinimalDev committed:
chore: implement redis cache for calendars api v2

ThyMinimalDev committed:
fixup! chore: implement redis cache for calendars api v2

ThyMinimalDev commented:
#23616

ThyMinimalDev committed:
fixup! fixup! chore: implement redis cache for calendars api v2

ThyMinimalDev commented on apps/api/v2/src/ee/calendars/services/calendars.service.ts:
```diff
@@ -187,6 +205,27 @@
+    this.redisService.del(REDIS_CALENDARS_CACHE_KEY(userId));
+  }
+
+  async getCalendarCache(userId: number) {
+    const cachedResult = (await this.redisService.get(
+      REDIS_CALENDARS_CACHE_KEY(userId)
+    )) as ConnectedDestinationCalendars;
+    return cachedResult;
+  }
+
+  async setCalendarCache(userId: number, calendars: ConnectedDestinationCalendars) {
+    await this.redisService.set(
+      REDIS_CALENDARS_CACHE_KEY(userId),
+      JSON.stringify(calendars),
+      { ttl: 10000 } // 10 sec
+    );
   }
```
no need to parse it it's done already within redisService.get

ThyMinimalDev committed:
fix leverage genetic type of redis.get

ThyMinimalDev committed:
Merge branch 'main' into chore-calendar-cache-api-v2

supalarry commented on apps/api/v2/src/ee/calendars/controllers/calendars.controller.ts:
```diff
@@ -83,7 +84,8 @@ export class CalendarsController {
     private readonly icsFeedService: IcsFeedService,
-    private readonly calendarsRepository: CalendarsRepository
+    private readonly calendarsRepository: CalendarsRepository,
+    private readonly redisService: RedisService
```
Do we need to inject `redisService` here if it's not used?

supalarry commented on apps/api/v2/src/ee/calendars/services/calendars.service.ts:
```diff
@@ -187,6 +205,27 @@ export class CalendarsService {
+    await this.redisService.del(REDIS_CALENDARS_CACHE_KEY(userId));
+  }
+
+  async getCalendarCache(userId: number) {
```
I would rename it `getConectedAndDestinationCalendarsCache` and also have `setConnectedAndDestinationCalendarsCache` because based on the awaited return type it's not clear what exactly is in the cache. It's a bit longer name but would inform what's cached + caching is called in destination and connected caledars services.

supalarry commented on apps/api/v2/src/ee/calendars/services/calendars.service.ts:
```diff
@@ -187,6 +205,27 @@ export class CalendarsService {
+    await this.deleteCalendarCache(userId);
+  }
+
+  async deleteCalendarCache(userId: number) {
```
Could we move the newly added cache functions to `CalendarsCacheService` and then re-use the cache service? Would help making this class smaller and separate responsibilities.

supalarry reviewed (commented):
In addition to comments below I think it would be useful to comment somewhere (possibly in the cache service) how the caching works - we invalidate cache if destination or connected caledars are added, updated or deleted and cache is _set only when fetching all connected and destination calendars_.

I am saying it because intuitively I would expect cache to be updated directly when a calendar is added or updated instead of invalidating all cache, but since we depend on `getConnectedDestinationCalendarsAndEnsureDefaultsInDb` I understand why we do what we do.

devin-ai-integration[bot] committed:
refactor: address supalarry's PR feedback

- Create CalendarsCacheService to separate cache responsibilities
- Rename cache methods to be more descriptive (getConnectedAndDestinationCalendarsCache, setConnectedAndDestinationCalendarsCache)
- Remove unused redisService injection from CalendarsController
- Move all cache-related functionality to dedicated service

Addresses comments 9, 10, and 11 from supalarry on PR #23622

Co-Authored-By: morgan@cal.com <morgan@cal.com>

supalarry reviewed (changes requested):
review:

devin-ai-integration[bot] committed:
refactor: address supalarry's latest PR feedback

- Remove redundant deleteConnectedAndDestinationCalendarsCache from CalendarsService
- Update controller to directly inject and use CalendarsCacheService
- Extract TTL constant and remove manual JSON.stringify in cache service

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into chore-calendar-cache-api-v2

devin-ai-integration[bot] committed:
fix: update IcsFeedService to use CalendarsCacheService directly

- Inject CalendarsCacheService into IcsFeedService constructor
- Replace calendarsService.deleteCalendarCache with calendarsCacheService.deleteConnectedAndDestinationCalendarsCache
- Completes the refactoring to eliminate all old cache method references

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: update destination-calendars and selected-calendars services to use CalendarsCacheService directly

- Inject CalendarsCacheService in both service constructors
- Replace calendarsService.deleteCalendarCache calls with calendarsCacheService.deleteConnectedAndDestinationCalendarsCache
- Add CalendarsCacheService to module providers for proper dependency injection
- Completes migration of all cache method references to use dedicated cache service

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
chore: use CalendarsCacheService

ThyMinimalDev committed:
Merge branch 'main' into chore-calendar-cache-api-v2

emrysal reviewed (dismissed):
yes plz ❤️

ThyMinimalDev committed:
update docs
